### PR TITLE
Fixes #3460 - Adds Intro to Progressive Web App Architecture

### LINKED ISSUE #3460: Intro to Progressive Web App Architecture
ILT content: https://docs.google.com/document/d/1gOVBluhK6MHIPSXkpA2MQMrdGYbJfBTowbHNYWc2Of0/edit#heading=h.wb3gqylk8pms

IIRC, we only have AppShell content up in web updates at the moment. It doesn't cover the same level of depth as this new write-up does.

### PR DESCRIPTION
**Pull request by addyosmani**
Heyo. This is a first-pass on PWA Architecture. I've tried to link to relevant /web docs where it made sense, take advantage of devsite styling for callouts and ensure `gulp test` passed. Images have all been run through ImageOptim.

Todos:
- [ ]  Tech writer review
- [ ]  Add Vivian to the contributors list and add her to the author list on this doc
- [ ]  Investigate other improvements I can make to the content

### CODE REVIEW

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 1]
```diff
@@ -0,0 +1,922 @@
...
+
+<img src="images/0.png" alt="After a period of idleness your service worker script is stopped automatically but when the next network request is made when the page is loaded again the service worker is started back up and can immediately respond to fetch events." class="screenshot"/>
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Based on the description you have, I'd recommend wrapping this in a figure and using the figcaption tag in addition to the alt tag.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 2]
```diff
@@ -0,0 +1,922 @@
...
+
+Note: Once for every given version of a service worker, the service worker is registered and activated and, if necessary, performs cache cleanup of any out-of-date resources that are no longer needed in your shell.
+
+**What about browsers that do not support service workers?**
+
+The app shell model is great but how does it work in browsers that do not support service workers? Don’t worry! Your web app can still be loaded even if a browser is used without service workers. Everything necessary to load your UI (e.g. HTML, CSS, JavaScript) is the same whether or not you use service workers. The service worker simply enables native-like features to your app like offline-caching. Without a service worker, your app continues to operate via HTTPS requests instead of cached assets. In other words, when service workers are not supported, the assets are not cached offline but the content is still fetched over the network and the user still gets a basic experience. 
+
+### Components {: #components }
+
+<table class="details">
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Nit - for two col tables, I strongly encourage making the table responsive.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 3]
```diff
@@ -0,0 +1,922 @@
...
+
+3. Server-side rendering full page (full page caching)
+
+Note: For browsers that do not support service workers, we gracefully degrade to still server-side rendering content (for example, iOS). 
+
+**Full client-side or server-side rendering the full page (full page caching, potential for JSON payload bootstrapping via server)**
+
+#### Table of Known Patterns for Building PWAs
+
+<table class="green">
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Try to not use colors, but instead use the DevSite styles. If there isn't a specific reason to use green here (ie, not details), then don't use green

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+
+## Migrating an Existing Site to PWA {: #migrating-an-existing-site-to-pwa }
+
+When migrating to a PWA, there are no hard and fast requirements around what to cache. You might, in fact, find it useful to think of your offline strategy as a series of milestones. It is feasible to begin by adding a simple service worker and just caching static assets, such as stylesheets and images, so these can be quickly loaded on repeat visits. 
+
+The next step might be caching the full-page HTML or caching the app shell to serve the empty UI first, and then allow the data layer to be pulled in from the server. This is the equivalent of a server sending down a rendered page without any results. 
+
+Each milestone allows you to deploy separately, measure the potential performance gains from each step you explore, and progressively roll out a better PWA. 
+
+Note: Understanding the network traffic is key to successful migration. You can use the guidelines in [Measure Resource Loading Times](https://developers.google.com/web/tools/chrome-devtools/profile/network-performance/resource-loading) to get started using the Network DevTools panel.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Don't use fully qualified URLs, instead use full urs. Specifically change the url here from https://developers.google.com/web/tools/... to /web/tools/...

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 5]
```diff
@@ -0,0 +1,922 @@
...
+
+### Migrating an Existing Site with Server Rendering to PWA
+
+Server-rendered pages can vary in complexity, either being (primarily) static HTML pages or involve more dynamic content. It is useful to think about how you might want to handle dynamic content as a number of different offline caching strategies can be used here. Jake Archibald's[ Offline Cookbook](/web/fundamentals/instant-and-offline/offline-cookbook/) is a good reference point once you moved your site over to [HTTPS](https://developers.google.com/web/fundamentals/security/encrypt-in-transit/why-https?hl=en), added a [Web App manifest](https://developers.google.com/web/updates/2014/11/Support-for-installable-web-apps-with-webapp-manifest-in-chrome-38-for-Android?hl=en) and can start crafting your service worker story.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Odd spacing in [ Offline Cookbook...

Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Web App Manifests should point to the manifest docs, not to an Update.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 6]
```diff
@@ -0,0 +1,922 @@
...
+
+Once you decide on a strategy for caching then you must implement it. A SPA architecture is often recommended when using an app shell, but it can take some time to refactor an existing site/app over to this architecture. If refactoring is a daunting task or if using an exclusively SSR approach is your only option for now, then you can still take advantage of service worker caching. But, you might end up treating your UI app shell the same way you would dynamic content. 
+
+* A cache-first strategy will not be entirely safe here if your server-rendered content is not entirely static and may change.
+
+* A [cache/network race](https://jakearchibald.com/2014/offline-cookbook/#cache-network-race) approach might work as with some combinations of hardware, getting resources from the network can be quicker than going to disk. Just keep in mind that requesting content from the network when the user has some copy of it on their device can waste potentially costly data.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Should point to within our own docs.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+
+* A [network-first approach that falls back to the cache](https://jakearchibald.com/2014/offline-cookbook/#network-falling-back-to-cache) might also work. Effectively, provide online users with the most up to date version of the content, but offline users get an older cached version. If a network request succeeds, then ensure the cached version gets updated.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Point within our own docs

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 8]
```diff
@@ -0,0 +1,922 @@
...
+
+* A [network-first approach that falls back to the cache](https://jakearchibald.com/2014/offline-cookbook/#network-falling-back-to-cache) might also work. Effectively, provide online users with the most up to date version of the content, but offline users get an older cached version. If a network request succeeds, then ensure the cached version gets updated. 
+
+Any of these strategies implements a web app that works offline. However, it is possible for data (any common HTML between /route1, /route2 , /route3, etc) to be cached twice. There can be performance and bandwidth hits when going to the network for the full content of the page as opposed to the app shell approach only fetches content (instead of content + UI). This can be mitigated using proper [HTTP browser caching headers](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching).
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: /route1, /route2 should be quoted in back ticks so they appear as code

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 9]
```diff
@@ -0,0 +1,922 @@
...
+
+### Additional Migration and Architectural Considerations {: #additional-considerations }
+
+* The app shell architecture comes with some challenges because the network request for content is delayed by the app shell loading from the cache, the JavaScript executing, and initiating the fetch. Eventually, Streams is a viable option in this case. Until then, the four known patterns described earlier for building PWAs are valid approaches.
+
+* The app shell should be managed with a cache-first strategy (cache-first, network-fallback). The point is to get reliable performance and to achieve this you must get the cache out of the network. See the [Table of Common Caching Strategies](#heading=h.j8pq42u9lupr) for more information.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: `#heading=h....`

Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: cache-first, network-fallback should link to docs

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 10]
```diff
@@ -0,0 +1,922 @@
...
+
+Assume you are building a simple blog reader. The components of a simple app shell include:
+
+* A link to the manifest file
+* Navigation UI and logic
+* The code to display posts after they are retrieved from the server (and store them in a local database)
+* The code to display comments (also storing them in the database)
+* Optionally, the code for posting comments
+
+Your index.html file in your work directory should look something like the following code. This is a subset of the actual contents and is not a complete index file. See [https://app-shell.appspot.com/](https://app-shell.appspot.com/) for a real-life look at a very simple app shell.
```
Comment (petele) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: index.html should be back-ticked

**Comment by addyosmani** [comment]
Thanks for the review, @petele! I'll get those comments addressed up. Per our thread, happy to either jump on a hangout to chat about where this fits or wait for your thoughts.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 11]
```diff
@@ -0,0 +1,922 @@
...
+
+Service worker caching should be considered a progressive enhancement. If your web app follows the model of conditionally registering a service worker only if it is supported, then you get offline support on browsers with service workers and on browsers that do not support service workers. The offline-specific code is never called and there is no overhead or breakage for older browsers. 
+
+
+### Key Concepts {: #key-concepts }
+
+The app shell approach relies on caching the "shell" of your web application using a service worker. Using the **_app shell + dynamic content model_** greatly improves app performance and works really well with service worker caching as a progressive enhancement. Progressively enhancing your web app means you can gradually add in features like offline caching, push notifications, and add-to-home-screen.
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: First sentence. I've heard this already.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 12]
```diff
@@ -0,0 +1,922 @@
...
+
+2. After the shell content loads, the app requests content to populate the view. The app shell plus dynamic content equals the compete rendered page. Next, the SPA requests content (for example, via [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) or the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)) and page content is fetched and used to populate the view. And, once those handlers complete, your service worker enters into an idle state. So, the service worker is running but it is not doing anything until a network request fires off a new event. And, in response to a network request, a fetch event handler intercepts the request and responds as you see fit. After a period of idleness your service worker script is stopped automatically but when the next network request is made when the page is loaded again the service worker is started back up and can immediately respond to fetch events.
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: "when the next network request is made when the page is loaded again" Is there a missing conjunction? Should it be "when the next network request is made _or_ when the page is loaded again"

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md (hunk 1 above)
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: I've made suggestions and edits for this text in the previous paragraph. Make sure they get copied.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 13]
```diff
@@ -0,0 +1,922 @@
...
+
+
+Note: Once for every given version of a service worker, the service worker is registered and activated and, if necessary, performs cache cleanup of any out-of-date resources that are no longer needed in your shell.
+
+**What about browsers that do not support service workers?**
+
+The app shell model is great but how does it work in browsers that do not support service workers? Don’t worry! Your web app can still be loaded even if a browser is used without service workers. Everything necessary to load your UI (e.g. HTML, CSS, JavaScript) is the same whether or not you use service workers. The service worker simply enables native-like features to your app like offline-caching. Without a service worker, your app continues to operate via HTTPS requests instead of cached assets. In other words, when service workers are not supported, the assets are not cached offline but the content is still fetched over the network and the user still gets a basic experience.
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Not only have I read all this before now, I'm getting anxious to see some code.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 14]
```diff
@@ -0,0 +1,922 @@
...
+  </tr>
+  <tr>
+    <td>cache</td>
+    <td>There are two types of cache: the automatic browser cache and program-controlled caches (<a href="https://developer.mozilla.org/en-US/docs/Web/API/Cache">Cache API</a> and <a href="https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API">IndexedDB</a>). 
+The browser cache is a temporary storage location on your computer for files downloaded by your browser to display websites. Files that are cached locally include any documents that make up a website, such as HTML files, CSS style sheets, JavaScript scripts, as well as graphic images and other multimedia content.
+The service worker also creates cache independent of the browser cache using the Cache Storage API or IndexedDB. The cache objects hold the same kinds of assets as a browser cache but make them offline accessible.The service worker provides these to enable offline support in browsers. This is referred to as "Cache (Storage) API" in this document.
+
+The cache objects are one tool you can use when building your app, but you must use them appropriately for each resource. Several caching strategies are described in Determine the Best Caching Strategy. </td>
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: There's a link missing.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 15]
```diff
@@ -0,0 +1,922 @@
...
+  </tr>
+  <tr>
+    <td>client-side rendering (CSR)</td>
+    <td>Client-side rendering means JavaScript running in the browser produces HTML (probably via templating). The benefit is you can update the screen instantly when the user clicks, rather than waiting a few hundred milliseconds at least while the server is contacted to ask what to display. Sites where you mostly navigate and view static content can get away with mostly server-side rendering. Any portion of a page that is animated or highly interactive (a draggable slider, a sortable table, a dropdown menu) almost certainly uses client-side rendering.</td>
+  </tr>
+  <tr>
+    <td>dynamic content
+</td>
+    <td>Dynamic content is all of the data, images, and other resources that your web app needs to function, but exists independently from your app shell. Although the app shell is intended to quickly populate the content of your site, users might expect dynamic content, in which case your app must fetch data specific to the user’s needs. Sometimes an app pulls this data from external, third-party APIs, and sometimes from first-party data that is dynamically generated or frequently updated.</td>
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: 'needs to function' sounds too much like the purpose of the app shell. Isn't this really the data/features users have come to see?

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 16]
```diff
@@ -0,0 +1,922 @@
...
+    <td>Fetch API</td>
+    <td>You can optionally use the <a href="https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API">Fetch API</a> to fetch resources. For example, if your web app is for a newspaper, it might make use of a first-party API to fetch recent articles, and a third-party API to fetch the current weather. Both of those types of requests fall into the category of dynamic content.</td>
+  </tr>
+  <tr>
+    <td>progressive enhancement</td>
+    <td>An approach to web development that begins with common browser features, and then adds in functionality or enhancements when the user’s browser supports more modern technologies.</td>
+  </tr>
+  <tr>
+    <td>PWA architecture styles</td>
+    <td>Any of several approaches to building PWAs based on the back-end technologies available and the performance requirements. The patterns include using an app shell, server-side rendering, client-side rendering, and others. These patterns are listed in Choose an Architecture.  </td>
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: Missing a link.

File: src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md [hunk 17]
```diff
@@ -0,0 +1,922 @@
...
+  </tr>
+  <tr>
+    <td>service worker</td>
+    <td>A type of web worker that runs alongside your web app but with a life span that is tied to the execution of the app’s events. Some of its services include a network proxy written in JavaScript that intercepts HTTP/HTTPS requests made from web pages. It also receives push messages. Additional features are planned in the future.</td>
+  </tr>
+  <tr>
+    <td>sw-precache</td>
```
Comment (jpmedley) [inline src/content/en/fundamentals/instant-and-offline/pwa-architecture/index.md]: You should probably clarify that these are tools to simplify implementation and not requirements.

**Pull request merged**